articles/realtime.md

```diff
@@ -0,0 +1,330 @@
+In general, the programmer should avoid page faults in the real-time code path.
+During a page fault, the CPU pauses all computation and loads the missing page from disk into RAM (or cache, or registers).
+Loading data from disk is a slow and unpredictable operation. However, page faults are necessary or else the computer will run out of memory.
+The solution is to avoid pagefaults.
+
+Dynamic memory allocation can cause poor real-time performance.
+Calls to `malloc/new` and `free/delete` will probably result in pagefaults.
+Additionally, the heap allocates and frees memory blocks in such a way that leads to memory fragmentation, which creates poor performance for reads and writes, since the OS may have to scan for an indeterminate amount of time for a free memory block.
+
+
+### Lock memory, prefault stack:
+
+```c
+    if (mlockall(MCL_CURRENT|MCL_FUTURE) == -1) {
+            perror("mlockall failed");
+            exit(-2);
+    }
+    unsigned char dummy[MAX_SAFE_STACK];
+
+    memset(dummy, 0, MAX_SAFE_STACK);
+```
+
+[`mlockall`](http://linux.die.net/man/2/mlockall) is a Linux system call for locking the process's virtual address space into RAM, preventing the memory that will be accessed by the process from getting paged into swap space.
+
+This code snippet, when run at the beginning of a thread's lifecycle, ensures that no pagefaults occur while the thread is running.
+`mlockall` locks the stack for the thread.
+The [`memset`](http://linux.die.net/man/3/memset) call pre-loads each block of memory of the stack into the cache, so that no pagefaults will occur when the stack is accessed. (3)
+
+### Allocate dynamic memory pool
+
+```c
+  if (mlockall(MCL_CURRENT | MCL_FUTURE))
+    perror("mlockall failed:");
+
+  /* Turn off malloc trimming.*/
+  mallopt(M_TRIM_THRESHOLD, -1);
+
+  /* Turn off mmap usage. */
+  mallopt(M_MMAP_MAX, 0);
+
+  page_size = sysconf(_SC_PAGESIZE);
+  buffer = malloc(SOMESIZE);
+
+  for (i=0; i < SOMESIZE; i+=page_size)
+  {
+      buffer[i] = 0;
+  }
+  free(buffer);
+```
+
+The intro to this section stated that dynamic memory allocation is bad. However, this code snippet shows how to make dynamic memory allocation real-time safe (mostly).
+It locks the virtual address space to a fixed size, disallows returning deallocated memory to the kernel via `sbrk`, and disables `mmap`.
+This effectively locks a pool of memory in the heap into RAM, which prevents page faults due to `malloc` and `free`. (3)
+
+Pros:
+
+* Can use malloc/new, free/delete, and even STL containers
+
+Cons:
+
+* Platform/implementation dependent
+* Must accurately predict bounded memory size for the process!
+* Using STL containers is therefore dangerous (unbounded sizes)
+* In practice, only works for processes with small memory footprint
+
+### Custom static allocators for STL containers
+
+To use the standard library safely, custom allocators for STL containers must be implemented that only allocate memory on the stack.
```

Comment:
thanks for the many in-depth resources. I'll link to this one since many of the finer points of memory allocation are probably out of the scope of this document, but definitely useful for a subtler understanding of safe dynamic memory allocation. And I'll reword this section.